File: tests/MicroserviceRgpd.TestDoubles/Ollama/a2_equivalence.py
"""Produit `a2-equivalence.json`, le jeu figé du contrôle d'équivalence du moteur A2 (#463).

Les vecteurs sont FABRIQUÉS, pas encodés : aucun Ollama ne tourne en test, et le contrôle ne porte
pas sur l'encodeur mais sur ce que le service fait des vecteurs — normalisation, régression, seuil,
prototype le plus proche. Chaque vecteur mêle un prototype de l'artefact, la direction de `coef` et
un bruit fixe, et la part de `coef` est ajustée par dichotomie pour viser un score. Les scores, les
décisions et les catégories sont ensuite calculés par `a2_model.py` lui-même, jamais par ce script.

Deux colonnes sont posées à ±1e-5 du seuil : un moteur dont le score s'écarte de plus de 1e-5 de
celui de Python rend au moins l'une des deux du mauvais côté.

Usage, depuis la racine du dépôt :

    uv run --with numpy python tests/MicroserviceRgpd.TestDoubles/Ollama/a2_equivalence.py \\
        <recherche_schema_v2>/src/pdmap/export \\
        src/MicroserviceRgpd.Infrastructure/Screenings/Embeddings/Artefact \\
        tests/MicroserviceRgpd.TestDoubles/Ollama/a2-equivalence.json
"""

from __future__ import annotations

import hashlib
import json
import sys
from pathlib import Path

import numpy as np

# (table, colonne, prototype visé ou None, écart visé au seuil)
COLUMNS = [
    ("users", "email", "contact", 0.25),
    ("users", "first_name", "identity", 0.30),
    ("users", "password_hash", "authentication", 0.20),
    ("patients", "blood_group", "health", 0.20),
    ("orders", "customer_id", "person_link", 0.15),
    ("products", "price", None, -0.30),
    ("products", "sku", None, -0.25),
    ("notes", "body", "free_text", 1e-5),
    ("orders", "status", "location", -1e-5),
]

# Ce que le double rend pour un texte qu'il ne connaît pas : une colonne sous le seuil.
DEFAULT_OFFSET = -0.35

# Aucun vecteur n'arrive normalisé : un moteur qui oublierait la normalisation L2 rendrait d'autres
# scores.
SCALE = 3.7


def main(module_dir: str, artefact_dir: str, output: str) -> None:
    sys.path.insert(0, module_dir)
    import a2_model  # noqa: PLC0415 — le module est recopié tel quel, hors de tout paquet

    model = a2_model.load(artefact_dir)
    rng = np.random.default_rng(463)
    coef_unit = model.coef / np.linalg.norm(model.coef)
    names = model.prototype_names

    def vector_for(prototype: str | None, offset: float) -> np.ndarray:
        noise = rng.standard_normal(model.coef.shape[0])
        noise /= np.linalg.norm(noise)
        base = (0.9 * model.prototypes[names.index(prototype)] + 0.35 * noise
                if prototype is not None else noise)
        target = model.threshold + offset

        def score(b: float) -> float:
            return float(model.score_vectors((base + b * coef_unit)[None, :])[0][0])

        low, high = -50.0, 50.0
        if not score(low) < target < score(high):
            raise SystemExit(f"score {target} hors d'atteinte pour {prototype}")
        for _ in range(200):
            middle = (low + high) / 2
            low, high = (middle, high) if score(middle) < target else (low, middle)
        return SCALE * (base + (low + high) / 2 * coef_unit)

    rows = []
    for table, column, prototype, offset in COLUMNS:
        vector = vector_for(prototype, offset)
        scores, categories, _ = model.score_vectors(vector[None, :])
        if prototype is not None and categories[0] != prototype:
            raise SystemExit(f"{table}.{column} : prototype le plus proche {categories[0]}, "
                             f"{prototype} visé")
        rows.append({
            "text": model.serialize(table, column),
            "table": table,
            "column": column,
            "score": float(scores[0]),
            "is_personal": bool(scores[0] >= model.threshold),
            "category": categories[0],
            "vector": [float(x) for x in vector],
        })

    default = vector_for(None, DEFAULT_OFFSET)
    default_score = float(model.score_vectors(default[None, :])[0][0])

    manifest = Path(artefact_dir, a2_model.MANIFEST_FILE).read_bytes()
    Path(output).write_text(json.dumps({
        "provenance": "vecteurs fabriqués par a2_equivalence.py ; scores, décisions et catégories "
                      "calculés par a2_model.py sur l'artefact a2_c1_logreg",
        "manifest_sha256": hashlib.sha256(manifest).hexdigest(),
        "threshold": model.threshold,
        "encoder": {"tag": model.encoder["tag"], "digest": model.encoder["digest"]},
        "columns": rows,
        "default": {
            "score": default_score,
            "is_personal": bool(default_score >= model.threshold),
            "vector": [float(x) for x in default],
        },
    }, ensure_ascii=False, indent=1) + "\n", encoding="utf-8")


if __name__ == "__main__":
    main(*sys.argv[1:4])

File: tests/MicroserviceRgpd.TestDoubles/Ollama/A2Equivalence.cs
using System.Text.Json;

namespace MicroserviceRgpd.TestDoubles.Ollama;

/// <summary>
/// Le jeu figé du contrôle d'équivalence : des colonnes, le vecteur que le double d'Ollama rend pour
/// chacune, et ce que <c>a2_model.py</c> en a décidé — score, décision, prototype le plus proche.
/// </summary>
/// <remarks>
/// <para>
/// ⚠️ <b>Les attendus sortent de Python, jamais du C#.</b> Le fichier est produit par
/// <c>a2_equivalence.py</c>, qui appelle <c>a2_model.py</c> sur l'artefact embarqué ; les recopier
/// depuis la sortie du moteur ferait de ce contrôle une tautologie.
/// </para>
/// <para>
/// ⚠️ <b>Les vecteurs sont fabriqués, pas encodés</b>, et c'est suffisant pour ce qu'on éprouve : ce
/// que le service fait d'un vecteur — normalisation, régression, seuil, prototype. Deux colonnes y
/// sont posées à ±1e-5 du seuil, si bien qu'un score qui s'écarterait davantage de celui de Python
/// rendrait l'une des deux du mauvais côté.
/// </para>
/// </remarks>
public static class A2Equivalence
{
  private static readonly Fixture Frozen = Load();

  /// <summary>Le seuil du manifest sur lequel Python a décidé.</summary>
  public static double Threshold => Frozen.Threshold;

  /// <summary>Le tag de l'encodeur que le manifest attend.</summary>
  public static string EncoderTag => Frozen.Encoder.Tag;

  /// <summary>Le digest de l'encodeur que le manifest attend, et que le double déclare servir.</summary>
  public static string EncoderDigest => Frozen.Encoder.Digest;

  /// <summary>L'empreinte SHA-256 du manifest sur lequel Python a calculé les attendus.</summary>
  public static string ManifestSha256 => Frozen.ManifestSha256;

  /// <summary>Les colonnes du jeu, avec ce que Python en a décidé.</summary>
  public static IReadOnlyList<FrozenColumn> Columns => Frozen.Columns;

  /// <summary>Le vecteur que le double rend pour ce texte — celui du jeu, ou le vecteur par défaut.</summary>
  public static IReadOnlyList<double> VectorOf(string text)
  {
    return Frozen.Columns.FirstOrDefault(column => column.Text == text)?.Vector ?? Frozen.Default.Vector;
  }

  private static Fixture Load()
  {
    using var stream = typeof(A2Equivalence).Assembly.GetManifestResourceStream("a2-equivalence.json")
      ?? throw new InvalidOperationException("Le jeu figé a2-equivalence.json n'est pas embarqué.");

    return JsonSerializer.Deserialize<Fixture>(stream, new JsonSerializerOptions(JsonSerializerDefaults.Web)
    {
      PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower,
    })!;
  }

  /// <summary>Une colonne du jeu.</summary>
  /// <param name="Table">Le nom de la table.</param>
  /// <param name="Column">Le nom de la colonne.</param>
  /// <param name="Text">Le texte au gabarit du manifest, tel que Python l'a sérialisé.</param>
  /// <param name="Score">Le score que Python a calculé — un attendu de test, jamais une sortie du service.</param>
  /// <param name="IsPersonal">La décision que Python a rendue.</param>
  /// <param name="Category">Le nom du prototype le plus proche, tel que le manifest l'écrit.</param>
  /// <param name="Vector">Le vecteur, non normalisé.</param>
  public sealed record FrozenColumn(
    string Table,
    string Column,
    string Text,
    double Score,
    bool IsPersonal,
    string Category,
    IReadOnlyList<double> Vector);

  private sealed record Fixture(
    string ManifestSha256,
    double Threshold,
    Encoder Encoder,
    IReadOnlyList<FrozenColumn> Columns,
    DefaultVector Default);

  private sealed record Encoder(string Tag, string Digest);

  private sealed record DefaultVector(double Score, bool IsPersonal, IReadOnlyList<double> Vector);
}
